src/helpers/StatsClient.ts

```diff
@@ -0,0 +1,176 @@
+	MICS_API_API_CALL_ERROR = 'mics_api_call_error',
+	MICS_API_API_CALL_RETRY = 'mics_api_call_retry',
+
+	AUDIENCE_FEED_BATCH_UPSERT = 'audience_feed_batch_upsert',
+	AUDIENCE_FEED_BATCH_DELETE = 'audience_feed_batch_delete',
+
+	AD_RENDER_BANNER_GENERATED_DEFAULT = 'ad_render_banner_generated_default',
+	AD_RENDER_BANNER_GENERATED_CUSTOM = 'ad_render_banner_generated_custom',
+
+	ACTIVITY_CREATED = 'activity_created',
+	ACTIVITY_FAILED_TO_CREATE = 'activity_failed_to_create',
+}
+
+export interface InitOptions {
+	/**
+	 * interval to send stats to datadog in ms (default = 10 minutes)
+	 */
+	timerInMs?: number;
+
+	/**
+	 * default plugins tags
+	 */
+	tags: Tags & { pluginType: PluginType; pluginName: string };
+
+	/**
+	 * An optional logger to send Metrics into logs (in debug mode)
+	 */
+	logger?: winston.Logger;
+}
+
+export interface IncrementToStatsOptions {
+	/**
+	 * @example
+	 * ```
+	 * name of your metrics and its value.
+	 * {processed_users: 4}
+	 */
+	metrics: Metrics;
+
+	/**
+	 * the main tag used to filter the metrics in datadog.
+	 * Ex: feedId
+	 */
+	scope?: string;
+}
+
+export interface Metrics {
+	[key: string]: number;
+}
+
+export interface TagsToScope {
+	/**
+	 * Custom tags retrived later on the plugin instantiation.
+	 * Ex: instance context infos.
+	 */
+	[scope: string]: Tags;
+}
+
+interface Stats {
+	[keyOrScope: string]: Metrics | number;
+}
+
+/**
+ * Send stats to datadog
+ */
+export class StatsClient {
+	private static instance: StatsClient;
+	private interval: NodeJS.Timer;
+	private stats: Stats;
+	private client: StatsD;
+	private logger?: winston.Logger;
+	private tagsToScope: TagsToScope;
+
+	private constructor(timerInMs: number, tags: Tags, logger?: winston.Logger) {
+		this.stats = {};
+		this.logger = logger;
+		this.tagsToScope = {};
+		this.client = new StatsD({
+			globalTags: {
+				workerId: process.env.PLUGIN_WORKER_ID || 'DEFAULT_PLUGIN_ID',
+				...tags,
+			},
+		});
+
+		if (!this.interval) {
+			this.interval = setInterval(() => this.sendStats(), timerInMs);
+		}
+	}
+
+	/**
+	 * @example
+	 * ```
+	 * private this.statsClient: StatsClient
+	 * constructor() {
+	 *   this.statsClient = StatsClient.init({ tags: { pluginType: PluginType.AUDIENCE_FEED, pluginName: 'MySuperPlugInName', segmentId: '1214'} });
+	 * }
+	 * ```
+	 */
+	static init({ timerInMs = 10 * 60 * 1000, tags, logger }: InitOptions): StatsClient {
+		return this.instance || (this.instance = new StatsClient(timerInMs, tags, logger));
+	}
+
+	/**
+	 * Increment some metrics
+	 * @example
+	 * ```
+	 * this.statClient.incrementToStats({scope: ctx.feedId, metrics: {processed_users: 4, users_with_mobile_id_count: 3}})
+	 * this.statClient.addToStats({scope: ctx.creativeId, metrics: {processed_users: 4}})
+	 * this.statClient.addToStats({metrics: {apiCallsSuccess: 4}})
+	 * this.statClient.addToStats({metrics: {apiCallsError: 1}})
+	 * ```
+	 */
+	public incrementToStats({ scope, metrics }: IncrementToStatsOptions): void {
```

Comment:
What is used the scope for ? 
```
      this.statsClient.addTagsToScope({
        scope: feedId,
        tags: { segmentId: segment.id, datamartId: segment.datamart_id },
      });
```
--> from the sample code
Is it another Tag ?

Reply:
It is your main tag if you add to choose one to be able to filter later.